src/jvmMain/kotlin/io/beatmaps/api/review.kt

```diff
@@ -491,14 +492,23 @@ fun Route.reviewRoute() {
                 reply.captcha,
                 {
                     val (insertedId, response) = newSuspendedTransaction {
-                        val allowedUsers = Review
-                            .join(Beatmap, JoinType.LEFT, Review.mapId, Beatmap.id)
-                            .select(Review.userId, Beatmap.uploader)
-                            .where { Review.id eq req.reviewId }
-                            .firstOrNull()
-                            ?.let {
-                                listOf(it[Review.userId].value, it[Beatmap.uploader].value)
-                            } ?: listOf()
+                        val review = Review.select(Review.mapId, Review.userId).where { Review.id eq req.reviewId and Review.deletedAt.isNull() }.singleOrNull()
+
+                        if (review == null) {
+                            return@newSuspendedTransaction Pair(null, ActionResponse(false, listOf("Review not found")))
+                        }
+
+                        val mapOfReview = Beatmap.joinCollaborators().selectAll().where { Beatmap.id eq review[Review.mapId] and Beatmap.deletedAt.isNull() }.complexToBeatmap().firstOrNull()
+
+                        if (mapOfReview == null) {
+                            return@newSuspendedTransaction Pair(null, ActionResponse(false, listOf("Map not found")))
+                        }
+
+                        val allowedUsers: List<Int> = listOf(
+                            mapOfReview.uploaderId.value,
+                            *mapOfReview.collaborators.values.map { it.id.value }.toTypedArray(),
+                            review[Review.userId].value
+                        )
```

Comment:
I'm wondering why the choice of these two queries...

- Option A: Take the existing query and join onto collaborators.
Seems simpler and you're only bringing back 3 columns so the cross-product effect isn't too bad.
- Option B: Just add a query to the collaborators table. No cross product at all.

But this way you're bringing back the whole map `selectAll()` 😱 and then throwing away most of the data...

Reply:
Yeah, this is not efficient. I was thinking the same - at the same time I don't expect this to be an actual issue performance wise.

Reason I did this was to use the complexToBeatmap function and have a list.

With Option A: If I just join collaborators, that might just be less efficient overall since joins are expensive and then I'd need to some odd, confusing parsing of the rows. So that didn't seem right.

Option B: Didn't consider that. So I'll probably do that :thinking: